### ISSUE TITLE
Can osDelay be emulated via getTimeStamp?

### ISSUE THREAD

**Issue by puzrin** [Ralim/usb-pd#9]
`PolicyEngine` already has `getTimeStamp`, but still requires `osDelay` for rare cases. Can `osDelay` be emulated via getTimeStamp internally?

Reasons: simplify use without RTOS.

**Comment by Ralim** [comment]
Could it be implemented that way: yes
Why isn't it: Don't want to lock up the thread for that long when using an RTOS. Also harder to ensure timing match if using extra wait states and ensuring the outer object is called often enough. Both are rare cases, so odds of running into it is low.
Also, almost all HAL's provide this exact function already, so its literally one extra param.

**Comment by puzrin** [comment]
How about constructor variants? If osDelay not passed => emulate

**Comment by Ralim** [comment]
Why though 😂 
Like, why should a _library_ be responsible for this?
Given this is primarily going to be used on small devices, the duplicated code doesnt feel wise :)

**Comment by puzrin** [comment]
Is library for "hardware first" or for "user first"? I can operate in any context :). But as OSS developper, wishing to make project popular, i'd prefer "user first" approach when possible.

> the duplicated code doesn't feel wise :)

IMO such statements are discuss-able only when metrics/bounds available. Without clear criterias talk will tend to fall into holywar without constructive output :)

**Comment by Ralim** [comment]
> Is library for "hardware first" or for "user first"? I can operate in any context :). But as OSS developper, wishing to make project popular, i'd prefer "user first" approach when possible.

Its user-first assuming user is at least competent in knowing roughly what they are setting up.
Its not user-first in the Arduino / PlatformIO way of usability>performance.


> IMO such statements are discuss-able only when metrics/bounds available. Without clear criterias talk will tend to fall into holywar without constructive output :)

Naturally, and to prove this point ill just save that my preference is smaller code size and to leave it up to the user to handle at their discretion. Given how trivial of a function it is in the simplest case, and how complex it can be in complex cases; that is a wide enough span to prefer to encourage the user to use their available functions.  :)

**Comment by puzrin** [comment]
> Its user-first assuming user is at least competent in knowing roughly what they are setting up.
> Its not user-first in the Arduino / PlatformIO way of usability>performance.

- Knowing how to install cmake / gcc / flash uploader is too much for beginners (especially on windows). It's more easy [for beginners] to install single vscode and let it organize the rest at project open.
- There are cases when you can ask "please change this line, upload on your device and see if problem fixed". Then user may be "not programmer" at all (but still able to install vscode).

I have nothing against your personal preferences/expectations, but IMO possible use is more wide.

**Comment by Ralim** [comment]
I completely agree with your statements, but if you are using a framework that is integrating this code, its _that_ framworks job to provide a good user experience for integrations.

The minimal required functions are [relatively workable](https://github.com/Ralim/STM32-USB-PD-Demo/blob/main/Workspace/USBPDDemo/Core/user/PDUser.cpp), while not abstracted away to the point of being useless for custom integrations.

I agree that example is quite long, but it _could_ be shortened easily, but prefer the most verbose example possible.



No-one is going to download vscode + this library and be able to flash it to a board. There is at least a good 2 more steps in the middle there. A lot of the work on installing this code, or how to hook to i2c or how to hook to delay is partly on those frameworks more so than this library.

**Comment by Seneral** [comment]
Just to add to the discussion, I'm looking to eventually implement this without a blocking function as well.
I'm not using a RTOS at all, so for me, it'd require quite a rewrite since I'd need the function to exit, let my code handle the waiting, and re-enter at a later time.
This is not super high priority for me at all, as the PD is currently setup at a time where a bit of blocking code isn't super harmful and I don't use PPS (yet).

**Comment by Ralim** [comment]
I'm open to a PR for a better api discussion on this, I dont have time to lead one at this point in time. My main focus is keeping code size as minimal as possible in my main usecase (over in IronOS where some devices have exceptionally little flash space left).

**Comment by puzrin** [comment]
https://github.com/pdsink/pdsink I decided to do "yet another sink" to solve all issues at once. It's not complete yet (no drivers), but the PE and PRL layers are carefully verified with the latest 3.2 spec and all available implementations. Internal APIs are event-based, and all platform dependencies are located in drivers only. Drivers can be done with protothreads (without RTOS), if needed.

It's not memory-optimized—that's doable, but out of priority now. The goal is to make a "solid" and easy-to-maintain solution for open-source projects, with all hardware support in mind (fusb302, stm32g0/4, wch).

**Comment by Ralim** [comment]
@puzrin btw forgot to loop back, but your library is looking very good these days. Hopefully one day I'll be able to replace this one with it so I have one less thing to maintain 🤣

**Comment by puzrin** [comment]
https://github.com/pdsink/pdsink/tree/master/src/drivers I've committed fusb302b driver draft (not tested). Have a couple of questions about your existing code:

- Seems you enable all interrupts in setup. This can cause an I_BC_LVL interrupt flood by BMC while receiving and transmitting. Known problem in all stacks, solved by complete/partial I_BC_LVL disable.
- Not clear why you use MDAC to check the VBUS level, instead of I_VBUSOK. Usually, manual measurement is required for SRC or with an external divider on the VBUS pin when existing hardware levels are insufficient.
- Suspicious delays
  - in chip setup (why??)
  - 10ms for levels measure (actually only 250uS required, if you use RTOS ticks - 2ms should guarantee as least 1ms after jitters)

Could you comment?

Also, for future "migration plans", if you have time, take a look at the flash/ram demands of my code. Let me know if those exceeded your expectations too much. I would keep that in mind for future plans.

**Comment by puzrin** [comment]
Thank you for valueable feedback. Right now, I'd suggest waiting for the MVP release, because I still need to stabilize API and squash history to keep it clean. Also, there is lack of examles/drives/docs.

I hope most of the past problems will go away because of honest PD spec implementation. But of course, if you have info about charges with problems - that will be very useful to check.

Here are some notes:

> I can add this myself, but for use in IronOS the firmware needs to be able to poll to check in negotiation has completed or not.

It may not be clear with existing code, but the general approach is "override DPM class" instead of constantly inventing new methods, to keep the API simple. DPM already has an interface to listen for events of PE toggles. The concrete wrapper depends on the specific needs of the concrete project. Existing DPM does not pretend to "have all you need" - it pretends to be flexible enough for any project.

> Must compile to same size or smaller than current code base (we have devices with approximately 0% free flash) when compiled for smaller targets (i.e. these can have no EPR support to get size down).

That will require some work. Currently, the top priority is to create the right architecture that is both usable/maintainable for 10+ years. I want to solve the problem of "pd sink for embedded" library "once and forever". That's why all optimizations were postponed until MVP release.

- `etl::fsm` uses runtime configuration (by etl design). Pinning states to FSM at compile time will require replacing the FSM implementation (I could not find better alternatives and postponed that for the future).
- Timestamps in timers are 64 bits.
- EPR support can be dropped, but that's not supported now.
- rx/tx buffers in core and drivers are duplicated, not shared.

Also note, since the library follows PD Spec very closely, that can take more code than in more "relaxed" alternatives.

> Needs to compile on your normal latest gcc compilers for arm and riscv via Make (I dont expect this to be an issue)

Core should work on any platform with `C++14` support (without `std`). Example with activated logs requires `C++17`. These demands can be reduced, but, in my opinion, are not necessary now.

The only problem can be in configuration without RTOS and without atomis support in the same time.

**Comment by Ralim** [comment]
Yeah definitely planning to wait until your happy with it before I dig in too deeply, just would love to make the PD stack someone else's problem tbh. Debugging PD is _not_ my idea of fun.

>But of course, if you have info about charges with problems - that will be very useful to check.

In general most are well behaved, but there are some that can be odd. Often the "most annoying" are laptops tbqh in terms of just spam of messages with short timeouts.

Even if I dont adopt using the library, absolutely happy to test it out at the least with all the stuff I have around here.


>It may not be clear with existing code, but the general approach is "override DPM class" instead of constantly inventing new methods, to keep the API simple. DPM already has an interface to listen for events of PE toggles. The concrete wrapper depends on the specific needs of the concrete project. Existing DPM does not pretend to "have all you need" - it pretends to be flexible enough for any project.

This is a good model 😁 


>Also note, since the library follows PD Spec very closely, that can take more code than in more "relaxed" alternatives.

Yeah I wouldn't be against a fork if I had to mask out code paths that we dont need. Same as some work may be required to ensure it compiles statically (no malloc/new) which I dont see as being something that is _your concern_ which is why I do expect if I changed to it, I'd have some work cut out for myself.

**Comment by puzrin** [comment]
I think everything will go well without the need to fork :). There are no dynamic allocations - only static objects are created at runtime start.

Do you have strict RAM limits or flash limits only? 1K RAM for vars/classes will be enough, without dirty optimizations and smelling code :).

What are the target devices where you have difficulties with flash size?

**Comment by puzrin** [comment]
It works! Finally :) . Still haven't squashed the commit history garbage.

I did a lot of refactoring and applied most of the optimizations mentioned above. Replacing etl::fsm with a custom one saved 10K flash and 2K memory, lol :). The only "big" missed thing - an option to build without PRL (without EPR) to save size. No objections in principle about this feature, but it's not a top priority now.

I'm interested in feedback. You can write any questions/wishes/complaints/ideas in the tracker. Current plan is:

- Test my bag of chargers and polish things. ~ 1 week.
- Integrate pdsink into my reflow table, to get feedback from using it in a real project. Several weeks.

This will get done regardless. But if more valuable feedback is available, I can make/plan adjustments.

**Comment by Ralim** [comment]
This is fantastic; and sorry for the delay. Life is a bit busy at the moment but I do plan to get to this and will let you know. Very hopeful I can move to using it sooner than later tbh.

>What are the target devices where you have difficulties with flash size?

TS80P is by far the worst, especially on CJK languages with lots of symbols in flash.

I have some wip ideas there to shake the tree and free up some more space but haven't been motivated to finish it.

RAM is _as far as I'm aware_ not the limited resource on most devices. Also dumping existing setup may free up about the same amount so shouldn't be crazy.


(I have a reminder set to get back to this, just cant give a timeline atm; sorry 😢 )
